sw/device/lib/dif/dif_plic.c

```diff
@@ -0,0 +1,317 @@
+#define PLIC_ALERT_HANDLER_LAST_IRQ kDifPlicAlertHandlerClassD
+#define PLIC_NMI_GEN_FIRST_IRQ kDifPlicNmiGenEsc0
+#define PLIC_NMI_GEN_LAST_IRQ kDifPlicNmiGenEsc3
+
+// Peripheral IRQ sources that interrupt the Ibex core
+#define PLIC_IBEX_SOURCE_FIRST_IRQ kDifPlicGpio0
+#define PLIC_IBEX_SOURCE_LAST_IRQ kDifPlicNmiGenEsc3
+
+// This define is used to calculate the IRQ index in IP, LE, IE registers.
+// These registers are 32bit wide, and in order to accommodate all the IRQs,
+// multiple of the same type registers are defined (IE00, IE01, ...). For
+// example, IRQ ID 33 corresponds to bit 1 in registers IP1, LE1, IE01. The
+// value 33 is chosen instead of 32 because IRQ IDs start from 1 rather than 0.
+#define PLIC_ID_TO_INDEX_MODULO 33u
+
+static bool plic_priority_validate(uint32_t priority) {
+  return priority <= RV_PLIC_MAX_PRIORITY;
+}
+
+/**
+ * Check the PLIC state for whether an IRQ has been claimed
+ */
+static bool plic_irq_claimed(dif_plic_t *plic) {
+  return (plic->claimed_irq.peripheral != kDifPlicPeripheralNone &&
+          plic->claimed_irq.source != kDifPlicNoIrq);
+}
+
+static void plic_irq_claimed_data_clear(dif_plic_t *plic) {
+  plic->claimed_irq.peripheral = kDifPlicPeripheralNone;
+  plic->claimed_irq.source = kDifPlicNoIrq;
+}
+
+/**
+ * Get IE, IP or LE register offset (IE00, IE01, ...) from IRQ source ID.
+ */
+static ptrdiff_t plic_offset_from_reg0(dif_plic_irq_src_id_t irq) {
+  uint8_t register_index = (uint32_t)irq / PLIC_ID_TO_INDEX_MODULO;
+  return register_index * sizeof(uint32_t);
+}
+
+/**
+ * Get PRIO register offset (PRIO0, PRIO1, ...) from IRQ source ID.
+ */
+static ptrdiff_t plic_priority_offset_from_reg0(dif_plic_irq_src_id_t irq) {
+  uint8_t register_index = (uint32_t)irq;
+  return register_index * sizeof(uint32_t);
+}
+
+/**
+ * Get IE, IP, LE register bit index from IRQ source ID. IRQ IDs start from 1,
+ * however IRQ bit index inside the registers start from 0. This means that in
+ * order to get IRQ bit index in the register, 1 has to be substracted from the
+ * obtained remainder.
+ */
+static uint8_t plic_bit_index_from_irq_id(dif_plic_irq_src_id_t irq) {
+  return ((uint32_t)irq % PLIC_ID_TO_INDEX_MODULO) - 1;
+}
+
+/**
+ * Get the target that will be interrupted by the IRQ source
+ */
+static bool plic_irq_target_from_src_id(dif_plic_irq_src_id_t irq,
+                                        dif_plic_target_t *target) {
+  if (irq >= PLIC_IBEX_SOURCE_FIRST_IRQ && irq <= PLIC_IBEX_SOURCE_LAST_IRQ) {
+    *target = kDifPlicTargetIbex0;
+  } else {
+    return false;
+  }
+
+  return true;
+}
+
+bool dif_plic_init(reg32_t base_addr, dif_plic_target_t target,
+                   dif_plic_t *plic) {
+  if (plic == NULL || base_addr.inner_ptr == 0) {
+    return false;
+  }
+
+  // Validate the requested target, and initialise the target specific registers
+  switch (target) {
+    case kDifPlicTargetIbex0:
+      reg32_write(base_addr, RV_PLIC_IE00, 0u);
```

Comment:
According to the spec these registers are already reset:
> After reset, RV_PLIC doesn’t generate any interrupts to any targets even if interrupt sources are set, as all priorities and thresholds are 0 by default and all IE values are 0.

Therefore, this code is not actually initializing anything.

Reply:
My remark was based on the comment above `dif_spi_device_init` (#1135), which may or may not be correct. While calling the `init` function more than once may be OK, IMHO it is more natural to provide a `reset` function since the `init` function expects a base address while `reset` function would look like a regular DIF that expects a `dif_plic_t`.